rust/kernel/debugfs.rs

```diff
@@ -0,0 +1,218 @@
+        // SAFETY: Calling a C function. `name` is a valid null-terminated
+        // string because it came from a [`CStr`] and `parent` is either null or
+        // valid because it came from a [`DebugFsDirectory`].
+        let dentry =
+            error::from_kernel_err_ptr(unsafe { bindings::debugfs_create_dir(name, parent_ptr) })?;
+        Ok(DebugFsDirectory { dentry, has_parent })
+    }
+}
+
+impl Drop for DebugFsDirectory {
+    fn drop(&mut self) {
+        // If this entry has a parent, we don't need to worry about removal
+        // because the parent will remove its children when dropped. Otherwise
+        // we need to clean up.
+        if !self.has_parent {
+            // SAFETY: Calling a C function. `dentry` must have been created by
+            // a call to `DebugFsDirectory::create` which always returns a
+            // valid `dentry`. There is no parent, so the
+            // `dentry` couldn't have been removed and must still be valid.
+            //
+            // This `dentry` and every `dentry` in it was created with either
+            // `DebugFsDirectory::create` or `DebugFsFile::create`. Both
+            // functions guarantee that the created `dentry` has a valide
+            // `inode` and the `inode`'s `i_private` field will be either null
+            // or come from calling `PointerWrapper::into_pointer` on a
+            // `Box<Box<dyn Any>>`. They both only create dentry's in debugfs.
+            // This makes it safe to call `remove_rust_dentry` on each `dentry`
+            // in `self.dentry`.
+            unsafe { debugfs_remove_with_callback(self.dentry, Some(remove_rust_dentry)) };
+        }
+    }
+}
+
+/// A `dentry` for a file in debugfs with a `T` stored in `i_private`.
+pub struct DebugFsFile<T> {
+    dentry: Option<*mut bindings::dentry>,
+    _t: PhantomData<T>,
+}
+
+// SAFETY: There are no public methods available on [`DebugFsFile`] so a thread
+// can't actually do anything with a `&DebugFsFile`. This makes it is safe to
+// share across threads.
+unsafe impl<T> Sync for DebugFsFile<T> {}
+
+impl<T: Any> DebugFsFile<T> {
+    /// Create a file in the `debugfs` directory under `parent`. If `parent` is
+    /// `None` then the file will be created at the root of the `debugfs`
+    /// directory and it will be removed on drop. If a `parent` is provided then
+    /// the `parent` is responsible for removing this file.
+    ///
+    /// # Safety
+    ///
+    /// `fops` must be valid when opening an `inode` with a `Box<Box<dyn
+    /// Any>>::into_pointer` that can be downcast to `T` stored in `i_private`.
+    unsafe fn create(
+        name: &CStr,
+        parent: Option<&mut DebugFsDirectory>,
+        data: T,
+        fops: &'static bindings::file_operations,
+    ) -> Result<DebugFsFile<T>> {
+        let has_parent = parent.is_some();
+        let name = name.as_char_ptr();
+        let boxed1: Box<dyn Any> = Box::try_new(data)?;
+        let boxed2 = Box::try_new(boxed1)?;
```

Comment:
Note that we're boxing twice here because `Box<dyn Any>` doesn't implement `PointerWrapper` since `dyn Any` is `!Sized`. I can look into making a `Box`-like `PointerWrapper` for unsized types if it would be generally useful.